### PR TITLE
*: enhancing group_concat function.

### PR DESCRIPTION
**Pull request by bobotu**
Only add `separator` currently.
Also add `order by` support in the parser, but doesn't implement it.

Originally wanted to add `order by` support together in this PR, but I found it is hard to implement with current implementation of executor and plan: `LogicalAggregation` pack all `aggFunc`  together, and iterate the input `Row`s and pass one `Row` to each `aggFunc`.

It seams like the only method to implement `order by` is let `concatFunction` cache all input rows, and sort all rows at last?

I make a variadic args list at `NewAggFunction` temporarily, because I am not sure how to support `order by` and there are too many call of this function. I want make it easy to change code in this PR, this variadic args list will change to exactly args at last.

### CODE REVIEW

**Comment by hanfei1991** [comment]
The algorithm of order by and group by is quiet complex. For example, concat(c1 order by c2) group by c1,c3.

1. You can require that the input is order by c1,c3,c2 or c3,c1,c2, so as to apply stream aggregation.
2. You can also require the input order by c1, c3 or c3, c1.
3. You can also require nothing and cache everything.

This should be decided by CBO Framework which I'm refactoring.

What if concat(s1 order by c1), concat (s2 order by c2) group by c3 ?

1. You can cache everything.
2. You can transform it to (concat(s1 order by c1), c3 group by c3) t1 join (concat (s2 order by c2), c3 group by c3) t2 on t1.c3 = t2.c3.

This should be decide by a transform-based CBO Framework which I'm planning to write.

In summary, it's very complex... Don't do it now.

File: expression/aggregation.go [hunk 1]
```diff
@@ -108,7 +110,14 @@ func NewAggFunction(funcType string, funcArgs []Expression, distinct bool) Aggre
 	case ast.AggFuncAvg:
 		return &avgFunction{aggFunction: newAggFunc(tp, funcArgs, distinct)}
 	case ast.AggFuncGroupConcat:
-		return &concatFunction{aggFunction: newAggFunc(tp, funcArgs, distinct)}
+		var sep string
+		for _, opt := range concatOpt {
```
Comment (shenli) [inline expression/aggregation.go]: Is it better to use  the first element of concatOpt as sep and the second element as orderby?  @coocood What's your opinion?

**Comment by hanfei1991** [comment]
I open an issue to record this question and close this pr temporily. I do very much appreciate your effort @bobotu
